File: 220.java
import java.util.TreeSet;

public class Solution {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        TreeSet<Integer> set = new TreeSet<>();
        if (nums == null || t < 0)
            return false;
        int n = nums.length;
        for (int i = 0; i < k; i++){
            if (set.contains(nums[i])) {
                return true;
            } else {
                Integer floor = set.floor(nums[i]);
                Integer ceil = set.ceiling(nums[i]);
                if (floor != null && floor >= nums[i] - t)
                    return true;
                else if (ceil != null && ceil <= nums[i] + t)
                    return true;
                set.add(nums[i]);
            }
        }
        for (int i = k; i < n; i ++) {
            if (set.contains(nums[i])) {
                return true;
            } else {
                Integer floor = set.floor(nums[i]);
                Integer ceil = set.ceiling(nums[i]);
                if (floor != null && floor >= nums[i] - t)
                    return true;
                else if (ceil != null && ceil <= nums[i] + t)
                    return true;
                set.add(nums[i]);
                set.remove(nums[i-k]);
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.containsNearbyAlmostDuplicate(new int[]{1, 2, 1}, 1, 0 ));
        System.out.println(solution.containsNearbyAlmostDuplicate(new int[]{1, 2, 1}, 2, 0 ));
        System.out.println(solution.containsNearbyAlmostDuplicate(new int[]{1, 9, 2}, 2, 0 ));
        System.out.println(solution.containsNearbyAlmostDuplicate(new int[]{1, 9, 2}, 2, 1 ));
    }
}

File: 220.py
class Solution(object):
    def containsNearbyAlmostDuplicate(self, nums, k, t):
        """
        :type nums: List[int]
        :type k: int
        :type t: int
        :rtype: bool
        """

        if t < 0 or nums == None or len(nums) == 0:
            return False
        data = {}
        d = t + 1
        for i, n in enumerate(nums):
            if i > k:
                del data[nums[i-k-1] / d]
            if n / d in data:
                return True
            elif n / d - 1 in data and abs(data[n/d - 1] - n) <= t:
                return True
            elif n / d + 1 in data and abs(data[n/d + 1] - n) <= t:
                return True
            data[n/d] = n

        return False

if __name__ == '__main__':
    test = Solution()
    print test.containsNearbyAlmostDuplicate([], 1, 0)
    print test.containsNearbyAlmostDuplicate([1,2,3], 1, 0)
    print test.containsNearbyAlmostDuplicate([1,2,1], 1, 0)
    print test.containsNearbyAlmostDuplicate([1,2,1], 2, 0)
    print test.containsNearbyAlmostDuplicate([1,9,2], 2, 0)
    print test.containsNearbyAlmostDuplicate([1,9,2], 2, 1)
    print test.containsNearbyAlmostDuplicate([-1,-1], 1, -1)
